### PR TITLE
[clang-tidy] Fix smart pointers handling in bugprone-use-after-move

### LINKED ISSUE #90174: [clang-tidy] bugprone-use-after-move  - false-negative
Example:

```
#include <memory>

struct A {};

struct ClassA
{
    ClassA(std::unique_ptr<A>);
};

class ClassB
{
    ClassB(std::unique_ptr<A> aaa)
        : aa(std::move(aaa))
    {
        a = std::make_unique<ClassA>(std::move(aaa));
    }
    std::unique_ptr<A> aa;
    std::unique_ptr<ClassA> a;
};

```
No issue from bugprone-use-after-move or from clang-analyzer-cplusplus.Move.

Example 2:
#include <iostream>
#include <memory>

struct Verbose
{
    Verbose() { std::clog << __PRETTY_FUNCTION__ << '\n'; }
    ~Verbose() { std::clog << __PRETTY_FUNCTION__ << '\n'; }
    Verbose(const Verbose&) { std::clog << __PRETTY_FUNCTION__ << '\n'; }
    Verbose(Verbose&&) { std::clog << __PRETTY_FUNCTION__ << '\n'; }
    Verbose& operator=(const Verbose&) { std::clog << __PRETTY_FUNCTION__ << '\n'; return *this; }
    Verbose& operator=(Verbose&&) { std::clog << __PRETTY_FUNCTION__ << '\n'; return *this; }

};

struct Test
{
    Test(std::unique_ptr<int> arg)
      : a(std::move(arg))
    {
        consume_again(std::move(arg));
    }
    void consume_again(std::unique_ptr<int>) {}
    std::unique_ptr<int> a;
};

int main() {
    std::unique_ptr<int> a = std::make_unique<int>(1);
    Test t1(std::move(a));
    Test t2(std::move(a));
}

Detected by clang-analyzer-cplusplus.Move, but not by bugprone-use-after-move

### PR DESCRIPTION
**Pull request by PiotrZSL**
Removed custom handling of smart pointers and added option IgnoreNonDerefSmartPtrs to restore
previous behavior if needed.

Closes #90174

### CODE REVIEW

**Comment by martinboehme** [comment]
More extensive comments on #90174 (let's continue the discussion there), but briefly:

> * Removed custom smart pointers handling (were hiding issues)

The standard smart pointers have a well-defined moved-from state (they are defined to be null after a move). Programmers may intentionally rely on this defined behavior. Warning on a (non-dereferencing) use-after-move for smart pointers as we do for other types would therefore result in false positives.

> * Changed 'move occurred here' note location to always point to 'std::move'

As discussed in more detail on #90174, the move doesn't actually occur in the `std::move()` (which is just a cast), and this can be important when sequencing rules come into play.

**Comment by PiotrZSL** [comment]
Rebase, bump

**Review by martinboehme — CHANGES_REQUESTED** [review]
Sorry - I only noticed now after your rebase that this had been sitting in my review queue  for quite a while.

*  I would suggest rewording the option name. The "ignore" (`IgnoreNonDerefSmartPtrs`) reads like a negation, which makes it harder to reason about what the option does. How about `AllowMovedSmartPtrUse`? (I think maybe the "non-deref" can be omitted from the option name -- I think it's sufficient that it's explained in the documentation).

*  In https://github.com/llvm/llvm-project/issues/90174, I suggested that if we wanted to change the default behavior, we should survey users to see what their opinions / preferences are. Have you done this? If not, I think the default behavior (when the option is not set) should remain unchanged.

**Comment by PiotrZSL** [comment]
@martinboehme 
_"I suggested that if we wanted to change the default behavior, we should survey users to see what their opinions / preferences are. Have you done this? If not, I think the default behavior (when the option is not set) should remain unchanged."_

Nope, there is no reliable way to "survey users". Default behavior of checks changes every release. Some checks are getting limited, by excluding some use cases, some checks are getting extended. In this case situation is easy. Users will update Clang-tidy, and if check will report "new issues", they have an option, to fix them or look into release notes to find out what changed in check and then change configuration to revert back to previous behavior. This is how things were handled in past for other checks.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/use-after-move.rst [hunk 1]
```diff
@@ -250,3 +251,14 @@ For example, if an additional member variable is added to ``S``, it is easy to
...
+
+Options
+-------
+
+.. option:: AllowMovedSmartPtrUse
+
+   If this option is set to `true`, the check will not warn about uses of
+   ``std::unique_ptr``, ``std::shared_ptr`` that are not dereferences. This
+   can be useful if you are using these smart pointers in a way that is not
+   idiomatic, but that you know is safe. Default is `false`.
```
Comment (martinboehme) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/use-after-move.rst:263]: "This can be useful if you are using these smart pointers in a way that is not idiomatic, but that you know is safe."

I think "not idiomatic" is too strong -- I would say this is a question of style. Maybe just leave out this entire sentence (i.e. describe only the behavior)?

File: clang-tools-extra/test/clang-tidy/checkers/bugprone/use-after-move-smart-ptr.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+void s(const std::unique_ptr<A> &);
+
+template <typename T, typename... Args> auto my_make_unique(Args &&...args) {
+  return std::unique_ptr<T>(new T(std::forward<Args>(args)...));
+}
+
+void natively(std::unique_ptr<A> x) {
+  std::unique_ptr<A> tmp = std::move(x);
+  std::unique_ptr<SinkA> y2{new SinkA(std::move(x))};
```
Comment (martinboehme) [inline clang-tools-extra/test/clang-tidy/checkers/bugprone/use-after-move-smart-ptr.cpp:32]: Is there a reason (here and possibly also below) that `SinkA` needs to be referenced via a `unique_ptr` instead of simply making it a local variable of `SinkA`?

I.e. why not simply do

```cxx
SinkA sink(std::move(x));
```

since the behavior we're interested in is what happens when we move `x` into the argument for the `SinkA` constructor, and not what happens with the `SinkA` itself.

(Also nit: I'm not clear on the reason for the `y2` variable name -- a) there is no `y1`, and b) maybe just `sink` to emphasize the type?)

**Review by martinboehme — CHANGES_REQUESTED** [review]

**Comment by martinboehme** [comment]
> Nope, there is no reliable way to "survey users". Default behavior of checks changes every release. Some checks are getting limited, by excluding some use cases, some checks are getting extended. In this case situation is easy. Users will update Clang-tidy, and if check will report "new issues", they have an option, to fix them or look into release notes to find out what changed in check and then change configuration to revert back to previous behavior. This is how things were handled in past for other checks.

OK, in that case this sounds fine.

**Review by 5chmidti — APPROVED** [review]
LGTM minus the other comments
